Sources/SmokeOperationsHTTP1Server/StandardHTTP1ResponseHandler.swift

```diff
@@ -119,6 +137,56 @@ public struct StandardHTTP1ResponseHandler<
+            let requestLatency = Date().timeIntervalSince(smokeInwardsRequestContext.requestStart).milliseconds
+            let serviceCallCount = smokeInwardsRequestContext.retriableOutputRequestRecords.count
+            let serviceCallLatency = smokeInwardsRequestContext.retriableOutputRequestRecords.reduce(0) { (retriableRequestSum, retriableRequestRecord) in
+                return retriableRequestSum + retriableRequestRecord.outputRequests.reduce(0) { (requestSum, requestRecord) in
+                    return requestSum + requestRecord.requestLatency.milliseconds
+                }
+            }
+            let retryWaitLatency = smokeInwardsRequestContext.retryAttemptRecords.reduce(0) { (retryWaitSum, retryAttemptRecord) in
+                return retryWaitSum + retryAttemptRecord.retryWait.milliseconds
+            }
+            let retriedServiceCalls = smokeInwardsRequestContext.retriableOutputRequestRecords.filter { requestRecord in
+                return requestRecord.outputRequests.count > 1
+            }
+            let serviceOnlyLatency = requestLatency - serviceCallLatency - retryWaitLatency
+            
+            var logComponents: [String] = []
+            
+            if serviceCallCount == 0 {
+                let logMessage = "Request completed in \(requestLatency) ms; (no service calls)."
+                logComponents.append("\(logMessage)")
+            } else {
+                let logMessage = "Request completed in \(requestLatency) ms; "
+                    + "\(serviceOnlyLatency) ms excluding service calls (there was \(serviceCallCount); "
+                    + "\(serviceCallLatency) ms service call latency, \(retryWaitLatency) ms retry backoff)."
+                logComponents.append("\(logMessage)")
+            }
+            
+            if retriedServiceCalls.count == 1 {
+                logComponents.append("1 outward service call was retried.")
+            } else {
+                logComponents.append("\(retriedServiceCalls.count) outward service calls were retried.")
+            }
+            
+            invocationContext.logger.info("\(logComponents.joined(separator: " "))")
+            
+            invocationContext.latencyTimer?.recordMilliseconds(requestLatency)
+            invocationContext.serviceLatencyTimer?.recordMilliseconds(serviceOnlyLatency)
+            invocationContext.outwardsServiceCallLatencySumTimer?.recordMilliseconds(serviceCallLatency)
+            invocationContext.outwardsServiceCallRetryWaitSumTimer?.recordMilliseconds(retryWaitLatency)
+            
+            if status.code >= 200 && status.code < 300 {
+                invocationContext.successCounter?.increment()
+            } else if status.code >= 400 && status.code < 500 {
+                invocationContext.failure4XXCounter?.increment()
+            } else {
+                invocationContext.failure5XXCounter?.increment()
+            }
```

Comment:
I think it's possible to aggregate these metrics into a single request to CloudWatch as long as it's using the same namespace. This will help reduce the overall latency.

Reply:
Its a valid comment although not a simple one to change as the swift-metrics API doesn't have a mechanism to aggregate metrics like this. I have made a change to the backend implementation in smoke-aws to put metrics into a queue and aggregate into a call to CloudWatch at regular intervals-

I have verified that these all get aggregated into the same call to CloudWatch.

https://github.com/amzn/smoke-aws/pull/79

Reply:
I think @tachyonics is right

Reply:
Gotcha, thanks @tachyonics